File: admin/src/lib/api/timeline.ts
import { api } from "./client";
import type { APIResponse } from "./types";

export interface TimelineTranslation {
  lang: string;
  title: string;
  description: string;
}

export interface TimelineMilestone {
  id: number;
  date: string;
  icon_name: string;
  sort_order: number;
  status: string;
  translations: TimelineTranslation[];
  created_at: string;
  updated_at: string;
}

export const timelineAPI = {
  list: (params?: Record<string, string>) =>
    api.get<APIResponse<TimelineMilestone[]>>("/timeline", params),

  get: (id: number | string) =>
    api.get<APIResponse<TimelineMilestone>>(`/timeline/${id}`),

  create: (data: Record<string, unknown>) =>
    api.post<APIResponse<TimelineMilestone>>("/timeline", data),

  update: (id: number | string, data: Record<string, unknown>) =>
    api.put<APIResponse<TimelineMilestone>>(`/timeline/${id}`, data),

  delete: (id: number | string) =>
    api.delete<APIResponse<null>>(`/timeline/${id}`),
};

File: bolg-api/internal/model/timeline.go
package model

import (
	"time"

	"gorm.io/gorm"
)

type TimelineMilestone struct {
	ID        uint           `gorm:"primaryKey;autoIncrement" json:"id"`
	Date      time.Time      `gorm:"type:date;index" json:"date"`
	IconName  string         `gorm:"type:varchar(50)" json:"icon_name"`
	SortOrder int            `gorm:"default:0;index" json:"sort_order"`
	Status    string         `gorm:"type:enum('draft','published');default:'published';index" json:"status"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	Translations []TimelineTranslation `gorm:"foreignKey:MilestoneID" json:"translations,omitempty"`
}

type TimelineTranslation struct {
	ID          uint   `gorm:"primaryKey;autoIncrement" json:"id"`
	MilestoneID uint   `gorm:"uniqueIndex:idx_timeline_trans;not null" json:"milestone_id"`
	Lang        string `gorm:"type:enum('en','zh');uniqueIndex:idx_timeline_trans;not null" json:"lang"`
	Title       string `gorm:"type:varchar(200);not null" json:"title"`
	Description string `gorm:"type:text" json:"description"`
}
